File: 2020/02/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
	"strings"
)

func validate1(min int, max int, char byte, password string) bool {
	count := 0
	for i := 0; i < len(password); i++ {
		if password[i] == char {
			count++
		}
	}
	return count >= min && count <= max
}

func validate2(x int, y int, char byte, password string) bool {
	return (password[x-1] == char) != (password[y-1] == char)
}

func main() {
	input, err := ioutil.ReadFile("2020/02/input.txt")
	if err != nil {
		panic(err)
	}
	answer1 := 0
	answer2 := 0
	lines := strings.Split(strings.TrimSpace(string(input)), "\n")
	delim := regexp.MustCompile("[- :]+")
	for _, line := range lines {
		values := delim.Split(line, 4)
		x, _ := strconv.Atoi(values[0])
		y, _ := strconv.Atoi(values[1])
		char := values[2][0]
		password := values[3]
		if validate1(x, y, char, password) {
			answer1++
		}
		if validate2(x, y, char, password) {
			answer2++
		}
	}
	fmt.Println("part 1:", answer1)
	fmt.Println("part 2:", answer2)
}

File: 2020/02/main.js
import { readFileSync } from "fs";

const input = readFileSync("2020/02/input.txt")
	.toString()
	.trim()
	.split("\n")
	.map((line) => {
		const values = line.split(/[- :]+/);
		return [parseInt(values[0]), parseInt(values[1]), values[2], values[3]];
	});

function validate1(min, max, char, password) {
	let count = 0;
	for (const c of password) {
		if (c === char) count++;
	}
	return count >= min && count <= max;
}

function validate2(x, y, char, password) {
	return (password[x - 1] === char) !== (password[y - 1] === char);
}

let answer1 = 0;
let answer2 = 0;
for (const row of input) {
	if (validate1(...row)) answer1++;
	if (validate2(...row)) answer2++;
}

console.log("part 1: " + answer1);
console.log("part 2: " + answer2);
